Add hands on code exemplifying basics (#152)
datacarpentry/image-processing: _episodes/02-image-basics.md

Viewed from a distance, these pixels seem to blend together to form the image
we see.

## Working with Pixels
As noted, in practice, real world images will typically be made up of a vast number of pixesl, and each of these pixels will be  one of potentially millions of colors. While we will deal with pictures of such complexity shortly, lets start our exploration with just 15 pixels in a 5 X 3 matrix with 2 colors and work our way up to that complexity, but first the necessary imports.

 ~~~
"""
 * Python libraries for learning and performing image processing.
 *
"""
import numpy as np
import skimage.io
import skimage.viewer
import matplotlib.pyplot as plt
~~~
{: .language-python}

Now that we have our libraries loaded, let's load our image data from disk.
 ~~~

> eight = np.loadtxt('data/eight.csv',delimiter=',')

Reviewer: Instead of using a csv file, perhaps we could generate an image file? So people don't get confused by yet another file format?

Author: I am onboard to try using an image file instead, but I am stymied.  Just changing the extension on my existing file doesn't seem to work.  I tried using plt.savefig() to output it, but I get a 640 X 480 X 4 matrix when I reimport that output.  I tried creating it and exporting from photoshop, and I got very close, but I cannot manage to get a matrix that outputs monochrome, I can only get greyscale with 255 for the white squares instead of 1.  Any ideas on how we can get a valid image file with only a 2d matrix with zeros and ones?

Author: Okay, seems like skimage.io.imsave can output the array in a format we can use if we use the tif extension.  Tried png, but that led to all kinds of warning and a false color 8 bit image display rather than monochrome.  Given how common tif is in medical applications, I assume we don't want to cut that one completely, so should we just proceed with reading it in as TIF image?

 ~~~
zero = np.loadtxt('data/eight.csv',delimiter=',')
zero[2,1]= 1.
zero_image = skimage.io.imshow(zero)
~~~
{: .language-python}
> ## Changing Pixel Values (5 min)
>
> Make another copy of eight named five, and then 
> change the value of pixels so you have a five in drawn in black pixels.
>
> > ## Solution
> >There are many possible solution, but one method would be . . .
> {: .solution}
{: .challenge}

## More colors
Up to now, we only had a 2 color matrix, but we can have more if we use other numbers or fractions. One common way is to use the numbers between 0 and 255 to allow for 256 different colors or 256 different levels of grey.  Let's try that out.
 ~~~
#make a copy of eight
three_colors = eight[:]

#multiply the whole matrix by 128
three_colors = three_colors * 128

# set the middle row (index 2) to the value of 255., so you end up with the values 0.,128.,and 255
three_colors[2,:] = 255.

> skimage.io.imshow(three_colors)

Reviewer: We might include the output of imshow in the lesson pages; but a few examples might help both instructors and learners while going through the steps, i.e., learners can cross-check their output with what is on the course page as a confirmation that "it works'.

Reviewer: > Add images for the code output.
Just saw that this is already covered.

Viewed from a distance, these pixels seem to blend together to form the image
we see.

## Working with Pixels

> As noted, in practice, real world images will typically be made up of a vast number of pixels, and each of these pixels will be  one of potentially millions of colors. While we will deal with pictures of such complexity shortly, lets start our exploration with just 15 pixels in a 5 X 3 matrix with 2 colors and work our way up to that complexity, but first the necessary imports.

Reviewer 2: ```suggestion
As noted, in practice, real world images will typically be made up of a vast number of pixels, and each of these pixels will be one of potentially millions of colors. While we will deal with pictures of such complexity shortly, let's start our exploration with 15 pixels in a 5 X 3 matrix with 2 colors and work our way up to that complexity, but first the necessary imports.
```

 ~~~
"""
 * Python libraries for learning and performing image processing.*
"""
import numpy as np
import skimage.io
import skimage.viewer
import matplotlib.pyplot as plt
import ipympl
~~~
{: .language-python}

> Now that we have our libraries loaded, we will run a Jupyter Magic Command that will insure our images display in our juptyer document with pixel information that will help us more efficiently run commands later in the session.

Reviewer 2: ```suggestion
Now that we have our libraries loaded, we will run a Jupyter Magic Command that will ensure our images display in our Jupyter document with pixel information that will help us more efficiently run commands later in the session.
```

 ~~~
%matplotlib widget
~~~
{: .language-python}

With that taken care of, let's load our image data from disk using the imread function from the skimage.io library and display it using the imshow function from the matplotlib library.
 ~~~
image = skimage.io.imread(fname="data/eight.tif")
plt.imshow(image)
~~~
{: .language-python}
![Image of 8](../fig/02-eight.png)

> You might be thinking, "That does look vaguely like an eight, and I see two colors but how can that be only 15 pixels". The display of the eight you see does use a lot more screen pixels to display our eight so large, but that doesn not mean there is information for all those screen pixels  in the file. All those extra pixels are a consequence of our viewer creating additional pixels through interpolation. It could have just displayed it as a tiny image using only 15 screen pixels if the viewer was written differently.

Reviewer 2: ```suggestion
You might be thinking, "That does look vaguely like an eight, and I see two colors but how can that be only 15 pixels". The display of the eight you see does use a lot more screen pixels to display our eight so large, but that does not mean there is information for all those screen pixels in the file. All those extra pixels are a consequence of our viewer creating additional pixels through interpolation. It could have just displayed it as a tiny image using only 15 screen pixels if the viewer was written differently.
```

While many image file formats contain descriptive metadata that can be essential, the bulk of a picture file is just arrays of numeric information that, when interpreted according to a certain rule set, become recognizable as an image to us. Our image of an eight is no exception, and skimage.io stored that image data in an array of arrays making a 5 x 3 matrix of 15 pixels. We can demonstrate that by calling on the shape property of our image variable and see the matrix by printing our image variable to the screen.

 ~~~
print(image.shape)
print(image)
~~~
{: .language-python}

> ![Image of 8 matrix](../fig/02-matrix-8.png)

Author: Great suggestion. The rationale was only lack of knowledge :) Will convert to output blocks and remove those images of matrices.

> While many imgage file formats contain descriptive metadata that can be essential, the bulk of a picture file is just arrays of numeric infomation that when interpreted according to a certain rule set becomes recognizable as an image to us. Our image of an eight is no exception, and skimage.io stored that image data in an array of arrays making a 5 x 3 matrix of 15 pixels. We can demonstrate that by calling on the shape properity of our image variable and see the matrix by printing our image variable to the screen.

Reviewer 2: ```suggestion
While many image file formats contain descriptive metadata that can be essential, the bulk of a picture file is just arrays of numeric information that, when interpreted according to a certain rule set, become recognizable as an image to us. Our image of an eight is no exception, and skimage.io stored that image data in an array of arrays making a 5 x 3 matrix of 15 pixels. We can demonstrate that by calling on the shape property of our image variable and see the matrix by printing our image variable to the screen.
```

>   Thus if we have tools that will allow us to manipulate these arrays of numbers, we can manipulate the image.  The numpy library can be particularly useful here, so let's try that out using numpy array slicing. Notice that the default behavior of the imageshow function appended row and an column numbers that will be helpful to us as we try to address indiviual or groups of pixels. First lets' load another copy of our eight, and then make it look like a zero.

Reviewer 2: ```suggestion
  Thus if we have tools that will allow us to manipulate these arrays of numbers, we can manipulate the image.  The numpy library can be particularly useful here, so let's try that out using numpy array slicing. Notice that the default behavior of the imageshow function appended row and column numbers that will be helpful to us as we try to address individual or groups of pixels. First let's load another copy of our eight, and then make it look like a zero.
```

> To make it look like a zero, we need to change the number underlying the center most pixel to be 1.  With the help of those row and column headers, at this small scale we can determine  the center pixel is in row labled 2 and column labeled 1. Using array slicing, we can then address and assign a new value to that position.

Reviewer 2: ```suggestion
To make it look like a zero, we need to change the number underlying the center most pixel to be 1.  With the help of those row and column headers, at this small scale we can determine the center pixel is in row labeled 2 and column labeled 1. Using array slicing, we can then address and assign a new value to that position.
```

 ~~~
zero = skimage.io.imread(fname="data/eight.tif")
zero[2,1]= 1.0
"""
The follwing line of code creates a new figure for imshow to use in displaying our output. Without it, plt.imshow() would overwrite our previous image in the cell above
""" 
fig, ax = plt.subplots()
plt.imshow(zero)
print(zero)
~~~
{: .language-python}

![Image of 0 matrix](../fig/02-matrix-0.png)

> ![Image of 0](../fig/02-zero.png)

Reviewer: This image appears in black and white. Perhaps just a slip, but I'd suggest to stick to same colormap for the entire section.

Reviewer 2: Agreed!

Author: implemented

>
>The most common manner to do this, and the one we will use in our programs,
>is to assign a modified Cartesian coordinate system to the image. The
>coordinate system we usually see in mathematics has a horizontal x-axis and
>a vertical y-axis, like this:
>
>![Cartesian coordinate system](../fig/01-cartesian.png)
>
>The modified coordinate system used for our images will have only positive
>coordinates, the origin will be in the upper left corner instead of the
>center, and y coordinate values will get larger as they go down instead of
>up, like this:
>
>![Image coordinate system](../fig/01-image-coordinates.png)
>
>This is called a *left-hand coordinate system*. If you hold your left hand
>in front of your face and point your thumb at the floor, your extended index
>finger will correspond to the x-axis while your thumb represents the y-axis.
>
>![Left-hand coordinate system](../fig/01-left-hand-coordinates.png)
>
>Until you have worked with images for a while, the most common mistake that
>you will make with coordinates is to forget that y coordinates get larger
>as they go down instead of up as in a normal Cartesian coordinate system.
{: .callout }

> > ## Changing Pixel Values (5 min)

Reviewer 2: ```suggestion

> ## Changing Pixel Values (5 min)
```

>
> Load another copy of eight named five, and then 

> > change the value of pixels so you have what looks like a 5 instead of 8. Dislay the image and print out the matrix as well.

Reviewer 2: ```suggestion
> change the value of pixels so you have what looks like a 5 instead of an 8. Display the image and print out the matrix as well.
```

>
> > ## Solution
> >There are many possible solutions, but one method would be . . .
> >~~~
> >five = skimage.io.imread(fname="data/eight.tif")
> >five[1,2]= 1.0
> >five[3,0]= 1.0
> >fig, ax = plt.subplots()
> >plt.imshow(five)
> >print(five)
> >~~~
> >{: .language-python}
> > ![Image of 5 matrix](../fig/02-matrix-5.png)
> > ![Image of 5](../fig/02-five.png)
> {: .solution}
{: .challenge}

## More colors
Up to now, we only had a 2 color matrix, but we can have more if we use other numbers or fractions. One common way is to use the numbers between 0 and 255 to allow for 256 different colors or 256 different levels of grey.  Let's try that out.
 ~~~
#make a copy of eight
three_colors = skimage.io.imread(fname="data/eight.tif")

#multiply the whole matrix by 128
three_colors = three_colors * 128

# set the middle row (index 2) to the value of 255., so you end up with the values 0.,128.,and 255
three_colors[2,:] = 255.
fig, ax = plt.subplots()
plt.imshow(three_colors)
print(three_colors)
~~~
{: .language-python}
![Image of three colors](../fig/02-three-colors.png)

> We now have 3 colors, but are they the three colors you expected?  They all appear to be on a continum of dark purple on the low end and yellow on the high end. This is a consequence of the default color map (cmap) in this library. You can think of a color map as an association or mapping of numbers to a specific color. However, the goal here is not to have one number for every color under the sun, but rather to have a continuum of colors that demonstrate relative intensity. In our specific case here for example, 255 or the highest intensity is mapped to yellow, and 0 or the lowest intensity is mapped to a dark purple. The best color map for your data will vary and there are many options built in, but this default selection was not arbitrary.  A lot of science went into making this the default due to its robustness when it comes to how the human mind interprets relative color values, grey-scale printability, and color-blind friendliness ([https://matplotlib.org/stable/tutorials/colors/colormaps.html](https://matplotlib.org/stable/tutorials/colors/colormaps.html), [https://bids.github.io/colormap/](https://bids.github.io/colormap/)).  Thus it is a good place to start, and you should change it only with purpose and forethought.  For now, let's see how you can do that using an alternative map you have likley seen before where it will be even easier to see it as a mapped continuum of intensities: greyscale.

Reviewer 2: ```suggestion
We now have 3 colors, but are they the three colors you expected?  They all appear to be on a continuum of dark purple on the low end and yellow on the high end. This is a consequence of the default color map (cmap) in this library. You can think of a color map as an association or mapping of numbers to a specific color. However, the goal here is not to have one number for every possible color, but rather to have a continuum of colors that demonstrate relative intensity. In our specific case here for example, 255 or the highest intensity is mapped to yellow, and 0 or the lowest intensity is mapped to a dark purple. The best color map for your data will vary and there are many options built in, but this default selection was not arbitrary.  A lot of science went into making this the default due to its robustness when it comes to how the human mind interprets relative color values, grey-scale printability, and color-blind friendliness ([https://matplotlib.org/stable/tutorials/colors/colormaps.html](https://matplotlib.org/stable/tutorials/colors/colormaps.html), [https://bids.github.io/colormap/](https://bids.github.io/colormap/)).  Thus it is a good place to start, and you should change it only with purpose and forethought.  For now, let's see how you can do that using an alternative map you have likely seen before where it will be even easier to see it as a mapped continuum of intensities: greyscale.
```

Author: Hey, Erin

Thanks for all the good feedback and suggestions.  I am working through them via the github interface, and nearly all them were easy to understand, but a couple of them puzzle me like this one.  Not sure if it is some artifact of the git or if I am just no seeing it.  This looks like no change to my eyes, so what am I missing if anything?

 
 ~~~
fig, ax = plt.subplots()
plt.imshow(three_colors,cmap=plt.cm.gray)
~~~
{: .language-python}
![Image in greyscale](../fig/02-grayscale.png)
Above we have exactly the same underying data matrix, but in greyscale.  Zero maps to black, 255 maps to white, and 128 maps to medium grey.

## Even More Colors

> This is all well and good at this scale, but what happens when we instead have a picture of a natural landscape that contains millions of colors.  Having a one to one mapping of number to color like this would be inefficient and make adjustments and building tools to do so very difficult.  Rather than larger numbers, the solution is to have more numbers in more dimensions. Let's see an example using a 4 X 4 matrix with 3 dimensions.  Rather than loading it from a file, we will generate an examples using numpy.

Reviewer: Perhaps a few sentences explaining channels would be helpful here for learners to see the rationale or using more dimensions?

Reviewer: Maybe a callout on color models?

Author: I think the start of the next chapter does a good job of exploring channels, so maybe it would be good to preview that as coming and maybe reference back to this in the start of the next chapter to connect those ideas

Author: Adjusted to this: This is all well and good at this scale, but what happens when we instead have a picture of a natural landscape that contains millions of colors. Having a one to one mapping of number to color like this would be inefficient and make adjustments and building tools to do so very difficult. Rather than larger numbers, the solution is to have more numbers in more dimensions. Storing the numbers in a multi-dimensional matrix where each color or property like transparency is associated with its own dimension allows for indiviudual contributions to a pixel to be adjusted independently. This ability to manipulate properties of groups of pixel separately will key to certain techniques explored in later chapters of this lesson. To get started let’s see an example of how different dimensions of information combine to produce a set of pixels using a 4 X 4 matrix with 3 dimensions for the colors red, green, and blue. Rather than loading it from a file, we will generate an examples using numpy.

~~~
#set the random seed so we all get the same matrix
pseudorandomizer = np.random.RandomState(2021)
#create a 4 X 4 checkerboard of random colors
checkerboard = pseudorandomizer.randint(0,255,size=(4,4,3)
                                       )
#restore the default map as you show the image
fig, ax = plt.subplots()
plt.imshow(checkerboard)
#display the arrays
print(checkerboard)
~~~
{: .language-python}

![Image of checkerboard matrix](../fig/02-matrix-checkerboard.png)
![Image of checkerboard](../fig/02-checkerboard.png)

> Previously we had one number being mapped to one color or intentsity. Now we are combining the effect of 3 numbers to arrive at a single color value.  Let's see an example of that using the blue square at the end of the second row, which has the index [1,3] 

Reviewer 2: ```suggestion
Previously we had one number being mapped to one color or intensity. Now we are combining the effect of 3 numbers to arrive at a single color value.  Let's see an example of that using the blue square at the end of the second row, which has the index [1,3]. 
```

~~~
# extract all the color information for the blue square
upper_right_square = checkerboard[1,3,:]
upper_right_square
~~~
{: .language-python}

This output: array([  7,   1, 110])

> The integers in order represent Red, Green, and Blue.  Looking at the 3 values and knowing how they map, can you understand why it is blue.  If we divide each value by 255, which is the maxium, we can determine how much it is contributing relative to its maximum potential. Effectively, the red is at 7/255 or 2.8 percent of its potential, the green is at 1/255 or 0.4 percent, and blue is 110/255 or 43.1 percent of its potential.  So when you mix those three intensities of color, blue is winning by a wide margin, but the red and green still contribute to make it a slightly different shade of blue than 0,0,110 would be on its own.

Reviewer 2: ```suggestion
The integers in order represent Red, Green, and Blue.  Looking at the 3 values and knowing how they map, can help us understand why it is blue. If we divide each value by 255, which is the maximum, we can determine how much it is contributing relative to its maximum potential. Effectively, the red is at 7/255 or 2.8 percent of its potential, the green is at 1/255 or 0.4 percent, and blue is 110/255 or 43.1 percent of its potential.  So when you mix those three intensities of color, blue is winning by a wide margin, but the red and green still contribute to make it a slightly different shade of blue than 0,0,110 would be on its own.
```

> These colors mapped to dimensions of the matrix may be referred to as channels.  It may be helpful to understanding if we display each of these channels independently.  We can do that by multiplying our image array representation with a 1d matrix that has a one for the channel we want to keep and zeros for the rest.

Reviewer 2: ```suggestion
These colors mapped to dimensions of the matrix may be referred to as channels.  It may be helpful to display each of these channels independently, to help us understand what is happening.  We can do that by multiplying our image array representation with a 1d matrix that has a one for the channel we want to keep and zeros for the rest.
```

~~~
red_channel = checkerboard * [1,0,0]
fig, ax = plt.subplots()
plt.imshow(red_channel)
~~~
{: .language-python}
![Image of red channel](../fig/02-red-channel.png)
~~~
green_channel = checkerboard * [0,1,0]
fig, ax = plt.subplots()
plt.imshow(green_channel)
~~~
{: .language-python}
![Image of green channel](../fig/02-green-channel.png)
~~~
blue_channel = checkerboard * [0,0,1]
fig, ax = plt.subplots()
plt.imshow(blue_channel)
~~~
{: .language-python}
![Image of blue channel](../fig/02-blue-channel.png)
If we look at the upper [1,3] square in all three figures, we can see each of those color contributions in action.  Notice that there are several squares in the blue figure that look even more intensely blue than square [1,3].  When all three channels are combined though, the blue light of those squares is being diluted by the relative strength of red and green being mixed in with them.

> 

Reviewer 2: ```suggestion
```

> 

Reviewer 2: ```suggestion
```

Author: This one was also a head scratcher for me. It seems like it is empty and got marked as outdated by github.  Do you know how to interpret the outdated label?

## 24 bit RGB Color

> This last color model we used, known as the *RGB (Red, Green, Blue)* model is the most common.

Reviewer 2: ```suggestion
This last color model we used, known as the *RGB (Red, Green, Blue)* model, is the most common.
```

## Image compression

> Before discussing additional formats, familiarity with image comression will be helpful.

Reviewer: ```suggestion
Before discussing additional formats, familiarity with image compression will be helpful.
```

> Let's delve into that subject with a challenge.  For that challenge, you will need to know about bits / bytes

Reviewer 2: ```suggestion
Let's delve into that subject with a challenge.  For this challenge, you will need to know about bits / bytes
```

and how those are use to express computer storage capacities.  If you already know, you can skip to the
challenge below.

>## Bits and bytes
>
>Before we talk specifically about images, we first need to understand how
>numbers are stored in a modern digital computer. When we think of a number, we
>do so using a *decimal*, or *base-10* place-value number system. For example, a
>number like 659 is 6 × 10<sup>2</sup> + 5 × 10<sup>1</sup> + 9 ×
>10<sup>0</sup>. Each digit in the number is multiplied by a power of 10, based
>on where it occurs, and there are 10 digits that can occur in each position
>(0, 1, 2, 3, 4, 5, 6, 7, 8, 9).
>
>In principle, computers could be constructed to represent numbers in exactly

> >the same way. But, as it turns out, the electronic circuits inside a computer

Reviewer 2: ```suggestion
>the same way. But, the electronic circuits inside a computer
```

>are much easier to construct if we restrict the numeric base to only two,

> >versus 10. (It is easier for circuitry to tell the difference between two

Reviewer 2: ```suggestion
>instead of 10. (It is easier for circuitry to tell the difference between two
```

> >voltage levels than it is to differentiate between 10 levels.) So, values in a

Reviewer 2: ```suggestion
>voltage levels than it is to differentiate among 10 levels.) So, values in a
```
